lib/Mongo/MongoGridFS.php

```diff
@@ -137,6 +245,95 @@ public function __get($id) {}
      * @param array $extra Other metadata to add to the file saved
      * @return mixed Returns the _id of the saved object
      */
-    public function put($filename, array $extra = array()) {}
+    public function put($filename, array $extra = array())
+    {
+        return $this->storeFile($filename, $extra);
+    }
+
+    private function ensureIndexes()
+    {
+        if ($this->ensureIndexes) {
+            return;
+        }
+        $this->ensureFilesIndex();
+        $this->ensureChunksIndex();
+        $this->ensuredIndexes = true;
+    }
+
+    private function ensureChunksIndex()
+    {
+        foreach ($this->chunks->getIndexInfo() as $index) {
+            if (isset($index['unique']) && $index['unique'] && $index['key'] === ['files_id' => 1, 'n' => 1]) {
+                return;
+            }
+        }
+        $this->chunks->createIndex(['files_id' => 1, 'n' => 1], ['unique' => true]);
+    }
+
+    private function ensureFilesIndex()
+    {
+        foreach ($this->getIndexInfo() as $index) {
+            if ($index['key'] === ['filename' => 1, 'uploadDate' => 1]) {
+                return;
+            }
+        }
+        $this->createIndex(['filename' => 1, 'uploadDate' => 1]);
+    }
+
+    private function insertChunksFromFile($file, $fileInfo)
+    {
+        $length = $fileInfo['length'];
+        $chunkSize = $fileInfo['chunkSize'];
+        $fileId = $fileInfo['_id'];
+        $offset = 0;
+        $i = 0;
+
+        while ($offset < $length) {
+            $data = fread($file, $chunkSize);
+            $this->insertChunk($fileId, $data, $i++);
+            $offset += $chunkSize;
+        }
+    }
+
+    private function calculateMD5($file, $length)
+    {
+        // XXX: this could be really a bad idea with big files...
+        $data = fread($file, $length);
```

Comment:
The file pointer may not be at the beginning, so you should seek to position 0 before reading from it. Furthermore, fread only reads up to `$length` bytes from the file. Since you don't know what's behind the stream you shouldn't count on fread returning all data at once.